### PR TITLE
add static analysis script to verify new feature gates are added as versioned feature specs

### LINKED ISSUE #125032: verification machinery for compatibility version
### What would you like to be added?

- [ ] verify that no new features are added through `utilfeature.DefaultMutableFeatureGate.Add` but should use `utilfeature.DefaultMutableFeatureGate.AddVersioned` instead.
- [ ]  verify `DefaultKubeEffectiveVersion` is up to date.
        * some script like verifying `DefaultKubeEffectiveVersion` is consistent with the release tag N for beta, and tolerates either N or N-1 with alpha

### Why is this needed?

To make sure https://github.com/kubernetes/enhancements/issues/4330 works properly, and not misused.

### PR DESCRIPTION
**Pull request by siyuanfoundation**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

As part of https://github.com/kubernetes/enhancements/issues/4330, we need to make sure any new features are added in the format of `featuregate.VersionedSpecs`.

The static analysis code analyzes the code and aggregate a list of all existing unversioned and versioned features.
* for verify, it compares the new list with the existing lists in test/static_analysis/test_data/unversioned_feature_list.yaml and test/static_analysis/test_data/versioned_feature_list.yaml, and verify fails when a feature is updated/deleted or a new feature is added.
* for update, it updates the lists in test/static_analysis/test_data/unversioned_feature_list.yaml and test/static_analysis/test_data/versioned_feature_list.yaml

Updates to the yaml files require approval from feature-approvers

#### Which issue(s) this PR fixes:

Partially Fixes https://github.com/kubernetes/kubernetes/issues/125032

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4330
```

### CODE REVIEW

File: hack/verify-test-featuregates.sh
Comment (jpbetz) [inline hack/verify-test-featuregates.sh:50]: nit: This doesn't seem to align with the name (and, presumably, the intended purpose) of this verify script since it was intended to verify _test_, where-as this verifies the use of non-test feature gates.

File: hack/verify-test-featuregates.sh
Comment (jpbetz) [inline hack/verify-test-featuregates.sh:48]: Can we always rely on git master being a clean representation of the tree minus what a contributor has changed?  Should we consider instead maintaining an allow list of old unversioned features and using that instead of diff-ing against master?

Comment (siyuanfoundation) [reply hack/verify-test-featuregates.sh:48]: I think the git master should be a clean representation of the tree if all new feature development goes into master. This is just more generic than maintaining multiple lists of old features for different components.

File: hack/verify-test-featuregates.sh
Comment (siyuanfoundation) [reply hack/verify-test-featuregates.sh:50]: The old https://github.com/kubernetes/kubernetes/blob/6a0aeb2adb8879dbdde7e06888aa91fed4526b8b/hack/verify-test-featuregates.sh#L46C45-L46C74 verifies non-test feature gates, so I just replace that one here. Maybe they should be moved to a new verify file?

File: hack/verify-test-featuregates.sh
Comment (jpbetz) [reply hack/verify-test-featuregates.sh]: Something I often do:

- Pull master
- Create branch A, do some development
- Decide I need to do something else for a while
- Checkout master
- Pull master
- Create branch B, do some development
- Find I need to finish something on branch A
- Checkout branch A
- (is master still a clean representation of branch A minus my development work?)

Comment (siyuanfoundation) [reply hack/verify-test-featuregates.sh]: when you 
* Checkout branch A for the 2nd time, the master could have stayed the same, or some old feature `FeatureA` have been removed. In case of master feature removal, you will see verify failure with the message of the `FeatureA` should not be added to the unversioned map. If you know you did not add `FeatureA` that should mean it is time to rebase. 

Verify should really be blocking if you are ready to merge. If you just want to grab someone else's branch to check it out, you can just ignore the verify, isn't it?

Comment (jpbetz) [reply hack/verify-test-featuregates.sh]: > If you know you did not add FeatureA that should mean it is time to rebase.

I expect this to be confusing to contributors.  I think we should avoid getting git involved in this.  We've had similar problems in the past with git comparisons and eventually migrated away from that approach to more hermetic/repeatable linter processes.

> Verify should really be blocking if you are ready to merge. If you just want to grab someone else's branch to check it out, you can just ignore the verify, isn't it?

Or maybe I'm going to use their branch for feature development... people use git in all sorts of ways..  Having a script that only works when people use git the way we expect them to use it seems prone to problems down the road.

**Review by logicalhan — COMMENTED** [review]
The way static analysis works for both conformance and metrics is that there are outputs of the static analysis, which create a file in an ACL'd directory. The verify script just ensures there is no diff between the committed file (easy to check, no weird git shenanigans required) and the presupposed regenerated output (note that this must be deterministic output). If there is a diff, there is an accompanying `update` script in the hack directory, which updates the committed file. Whenever that file changes, a set of approvers must review those changes. 

The way I imagined it would work here is that we would detect for all signatures of the the old feature gate methods and output those in a file, if we detect changes we fail the verify script, if there is a diff, you must update the file and be reviewed (only deletions should be allowed, no additions). This would eventually no-opt once we get rid of all usages.

**Comment by jpbetz** [comment]
> The way static analysis works for both conformance and metrics is that there are outputs of the static analysis, which create a file in an ACL'd directory. The verify script just ensures there is no diff between the committed file (easy to check, no weird git shenanigans required) and the presupposed regenerated output (note that this must be deterministic output). If there is a diff, there is an accompanying `update` script in the hack directory, which updates the committed file. Whenever that file changes, a set of approvers must review those changes.
> 
> The way I imagined it would work here is that we would detect for all signatures of the the old feature gate methods and output those in a file, if we detect changes we fail the verify script, if there is a diff, you must update the file and be reviewed (only deletions should be allowed, no additions). This would eventually no-opt once we get rid of all usages.

Yep, in this particular case I was expecting an allow list file of "not yet valid" cases to exist (in some ACL'd directory) and each time the script runs, as long as the allow list is unchanged, it passes.  If the allow list gets shorter, great!  If it get's longer, we wouldn't approve without a very good reason.

**Comment by siyuanfoundation** [comment]
> > The way static analysis works for both conformance and metrics is that there are outputs of the static analysis, which create a file in an ACL'd directory. The verify script just ensures there is no diff between the committed file (easy to check, no weird git shenanigans required) and the presupposed regenerated output (note that this must be deterministic output). If there is a diff, there is an accompanying `update` script in the hack directory, which updates the committed file. Whenever that file changes, a set of approvers must review those changes.
> > The way I imagined it would work here is that we would detect for all signatures of the the old feature gate methods and output those in a file, if we detect changes we fail the verify script, if there is a diff, you must update the file and be reviewed (only deletions should be allowed, no additions). This would eventually no-opt once we get rid of all usages.
> 
> Yep, in this particular case I was expecting an allow list file of "not yet valid" cases to exist (in some ACL'd directory) and each time the script runs, as long as the allow list is unchanged, it passes. If the allow list gets shorter, great! If it get's longer, we wouldn't approve without a very good reason.

Good points. I will work on the generation of the allow list and update the PR.

File: test/static_analysis/OWNERS
Comment (siyuanfoundation) [reply test/static_analysis/OWNERS:4]: Added api-machinery

My hope is other static analysis code can go in to this directory too, not just for feature gates. Is it ok to include sig-instrumentation in here?

Comment (logicalhan) [reply test/static_analysis/OWNERS:4]: Sure.

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

**Comment by BenTheElder** [comment]
This appears to add ~2m20s (out of ~49m spent on the main verify job), that's not ... terrible, but is it possible to implement this as a golangci-lint linter? That allows us to use a shared cache for loading and parsing the source tree versus other go linters.

Most of our static analysis tools are, excluding typecheck which predates golangci-lint and would actually be more expensive as part of golangci-lint (because the whole point is to check the source tree for multiple build target configurations)

**Comment by BenTheElder** [comment]
https://github.com/kubernetes/kubernetes/blob/0fc167103189a4d462c3cc7a17b360c3d998f4bf/hack/tools/tools.go#L28 is one example https://github.com/kubernetes-sigs/logtools/tree/main/logcheck

**Comment by logicalhan** [comment]
> This appears to add ~2m20s (out of ~49m spent on the main verify job), that's not ... terrible, but is it possible to implement this as a golangci-lint linter? That allows us to use a shared cache for loading and parsing the source tree versus other go linters.
> 
> Most of our static analysis tools are, excluding typecheck which predates golangci-lint and would actually be more expensive as part of golangci-lint (because the whole point is to check the source tree for multiple build target configurations)

Sounds like a potential improvement but I wouldn't block on it. We do need the safeguard and this is already an established pattern of statically analyzing k8s/k8s.

**Comment by pohly** [comment]
> Sounds like a potential improvement but I wouldn't block on it. We do need the safeguard and this is already an established pattern of statically analyzing k8s/k8s.

What was okay in the past might not be okay anymore when there is a better way to do it now...

If we accept this separate linter, who will update it to run as a golangci-lint plugin and when?

File: test/static_analysis/OWNERS
Comment (pohly) [reply test/static_analysis/OWNERS:4]: Is the plan that this `static_analysis` binary should get extended by other SIGs, i.e. is it meant to be a common framework for running static analysis tools?

A big :-1: from me for that. We have golangci-lint for that purpose.

If you absolutely don't want to integrate with that, then put this somewhere else under a different name. `cmd/featuregatecheck` seems more appropriate to me.

**Comment by BenTheElder** [comment]
> Sounds like a potential improvement but I wouldn't block on it. 

I am requesting that we consider this in advance while the tool is still small, and seeking input from other leads.

> We do need the safeguard [...]

Agree.

> [...] and this is already an established pattern of statically analyzing k8s/k8s.

Somewhat? 

Codegen is a bit different, as is the cross-platform typecheck, the other verify scripts are either not analyzing go at all (e.g. shellcheck) or don't actually load the go source tree (module file checks, etc). 

It would be preferable to avoid adding more things to presubmit that independently load the repo's go AST for a repo this large. We have a robust framework for static analysis already in hack/tools golangci-lint that has standardization and performance wins.

I think verify-importboss is probably the closest thing that exists currently that isn't golangci-lint. We probably wouldn't take the same approach today, import-boss is 9 years old and didn't have access to the tooling we have now.

**Comment by logicalhan** [comment]
> If we accept this separate linter, who will update it to run as a golangci-lint plugin and when?

It can be beta criteria for the KEP.

File: test/static_analysis/OWNERS
Comment (logicalhan) [reply test/static_analysis/OWNERS:4]: I'm okay with renaming the directory.

**Comment by BenTheElder** [comment]
> It can be beta criteria for the KEP.

That sounds fair to me. Along with https://github.com/kubernetes/kubernetes/pull/125830#discussion_r1698875677

I also do think we should prefer to get coverage in this release, for the reasons outlined in https://github.com/kubernetes/kubernetes/issues/125032#issue-2308889790

/priority important-soon

**Comment by pohly** [comment]
I'm okay with renaming the current `test/static_analysis` directory (ideally now before merging), then integrating it into golangci-lint later.

**Comment by dims** [comment]
> I'm okay with renaming the current test/static_analysis directory (ideally now before merging), then integrating it into golangci-lint later.

+1

**Pull request merged**
